cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -1425,31 +1431,86 @@ class AsofJoinNode : public ExecNode {
     // The reason for this is that there are cases at the end of a table where we don't
     // know whether the RHS of the join is up-to-date until we know that the table is
     // finished.
-    process_.Push(true);
+    PushProcess(true);
+
     return Status::OK();
   }
+  void PushProcess(bool value) {
+#ifdef ARROW_ENABLE_THREADING
+    process_.Push(value);
+#else
+    if (value) {
+      ProcessNonThreaded();
+    }
+#endif
+  }
 
-  Status StartProducing() override {
 #ifndef ARROW_ENABLE_THREADING
-    return Status::NotImplemented("ASOF join requires threading enabled");
+  bool ProcessNonThreaded() {
+    while (!process_task_.is_finished()) {
```

Comment:
How does it exit this loop? Seems  `process_task_` is only marked finished in `EndFromSingleThread` which won't be called if everything goes well.

Reply:
I think (but it's been a while) that `ProcessInner` will return `nullptr` if there is no more work to be done.  Then on line `1455` we have:

```
if (!out_rb) break;
```

Reply:
yep.